src/xrpld/app/misc/LendingHelpers.h

```diff
@@ -0,0 +1,444 @@
+    // the precision of the asset and loan.
+    Number periodicPayment;
+
+    // The total amount the borrower will pay over the life of the loan.
+    // Equal to periodicPayment * paymentsRemaining.
+    // This includes principal, interest, and management fees.
+    Number totalValueOutstanding;
+
+    // The total management fee that will be paid to the broker over the
+    // loan's lifetime. This is a percentage of the total interest (gross)
+    // as specified by the broker's management fee rate.
+    Number managementFeeOwedToBroker;
+
+    // The scale (decimal places) used for rounding all loan amounts.
+    // This is the maximum of:
+    // - The asset's native scale
+    // - A minimum scale required to represent the periodic payment accurately
+    // All loan state values (principal, interest, fees) are rounded to this
+    // scale.
+    std::int32_t loanScale;
+
+    // The principal portion of the first payment.
+    Number firstPaymentPrincipal;
+};
+
+/** This structure captures the parts of a loan state.
+ *
+ *  Whether the values are raw (unrounded) or rounded will depend on how it was
+ * computed.
+ *
+ *  Many of the fields can be derived from each other, but they're all provided
+ *  here to reduce code duplication and possible mistakes.
+ *   e.g.
+ *     * interestOutstanding = valueOutstanding - principalOutstanding
+ *     * interestDue = interestOutstanding - managementFeeDue
+ */
+struct LoanState
+{
+    // Total value still due to be paid by the borrower.
+    Number valueOutstanding;
+    // Principal still due to be paid by the borrower.
+    Number principalOutstanding;
+    // Interest still due to be paid to the Vault.
+    // This is a portion of interestOutstanding
+    Number interestDue;
+    // Management fee still due to be paid to the broker.
+    // This is a portion of interestOutstanding
+    Number managementFeeDue;
+
+    // Interest still due to be paid by the borrower.
+    Number
+    interestOutstanding() const
+    {
+        XRPL_ASSERT_PARTS(
+            interestDue + managementFeeDue ==
+                valueOutstanding - principalOutstanding,
+            "ripple::LoanState::interestOutstanding",
+            "other values add up correctly");
+        return interestDue + managementFeeDue;
+    }
+};
+
+// Some values get re-rounded to the vault scale any time they are adjusted. In
+// addition, they are prevented from ever going below zero. This helps avoid
+// accumulated rounding errors and leftover dust amounts.
+template <class NumberProxy>
+void
+adjustImpreciseNumber(
+    NumberProxy value,
+    Number const& adjustment,
+    Asset const& asset,
+    int vaultScale)
+{
+    value = roundToAsset(asset, value + adjustment, vaultScale);
+
+    if (*value < beast::zero)
+        value = 0;
+}
+
+inline int
+getVaultScale(SLE::const_ref vaultSle)
+{
+    if (!vaultSle)
+        return Number::minExponent - 1;  // LCOV_EXCL_LINE
+    return vaultSle->at(sfAssetsTotal).exponent();
```

Comment:
if the asset type is integral (XRP or MPT), would the exponent of `sfAssetsTotal` would always be 0? if so, should there be an invariant for this?